client/consensus/manual-seal/src/heartbeat_stream.rs

```diff
@@ -0,0 +1,133 @@
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! HeartbeatStream type and HeartbeatOptions used in instant seal
+use std::{pin::Pin};
+use futures::{
+	prelude::*,
+	task::{Context, Poll},
+};
+use tokio::time::{Duration, Instant, Delay};
+
+use crate::rpc::{EngineCommand};
+
+/// Constants for generating default heartbeat options
+const DEFAULT_MAX_BLOCKTIME: u64 = 30;
+const DEFAULT_MIN_BLOCKTIME: u64 = 1;
+const DEFAULT_FINALIZE: bool = false;
+
+/// Heartbeat options to manage the behavior of the heartbeat stream
+pub struct HeartbeatOptions {
+	/// The minimum amount of time to pass before generating a new block, in sec.
+	pub min_blocktime: u64,
+	/// The amount of time passed that a new heartbeat block will be generated, in sec.
+	pub max_blocktime: u64,
+	/// Control whether the generated block is finalized
+	pub finalize: bool,
+}
+
+impl Default for HeartbeatOptions {
+	fn default() -> Self {
+		Self {
+			min_blocktime: DEFAULT_MIN_BLOCKTIME,
+			max_blocktime: DEFAULT_MAX_BLOCKTIME,
+			finalize: DEFAULT_FINALIZE,
+		}
+	}
+}
+
+/// HeartbeatStream is combining transaction pool notification stream and generate a new block
+///   when a certain time has passed without any transactions.
+pub struct HeartbeatStream<Hash> {
+	/// The transaction pool notification stream
+	pool_stream: Box<dyn Stream<Item = EngineCommand<Hash>> + Unpin + Send>,
+	/// Future to control when the next block should be generated
+	delay_future: Delay,
+	/// To remember when the last block is generated
+	last_blocktime: Option<Instant>,
+	/// Heartbeat options
+	opts: HeartbeatOptions,
+}
+
+impl<Hash> HeartbeatStream<Hash> {
+	pub fn new(
+		pool_stream: Box<dyn Stream<Item = EngineCommand<Hash>> + Unpin + Send>,
+		opts: HeartbeatOptions
+	) -> Self {
+		if opts.min_blocktime > opts.max_blocktime {
+			panic!("Heartbeat options `min_blocktime` value must not be larger than `max_blocktime` value.");
+		}
+		Self {
+			pool_stream,
+			delay_future: tokio::time::delay_for(Duration::from_secs(opts.max_blocktime)),
+			last_blocktime: None,
+			opts
+		}
+	}
+}
+
+impl<Hash> Stream for HeartbeatStream<Hash> {
+
+	type Item = EngineCommand<Hash>;
+
+	fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
```

Comment:
The logic in here is confusing.

When I understood it correctly, you always only want to generate a block at a rate of `min_blocktime` and at maximum `max_blocktime` between blocks.

So what I would propose the following:

1. You always poll `delay_future` first. 
2. If the `delay_future` fired, check if the `pool_stream` had any events, if there was an event build a new block.
3. If there was no event, check if the `maximum_block` time passed and build a block if so.

`delay_future` should be changed to an interval to make it easier and you should also introduce an interval for the maximum block time.